### PR TITLE
Stop build on failure

### PR DESCRIPTION
**Pull request by katsadim**
Some of the most popular jenkins plugins offer by default the possibility to stop the build if something goes wrong. Namely:
* ArchiveArtifact: Fails the build if it finds no artifacts
* junit: Fails the build on empty test results
* publishHTML: Fails the build if it finds no html reports

This functionality is now introduced to cucumber-report-plugin and it could be helpful in a number of use cases. One of them is when we need a number of cucumber scenarios to pass before Jenkins is allowed to proceed with a deploy to production environment.

In order to maintain backwards compatibility with existing setups that do not expect the build to be stopped, user should explicitly tick the `Stop Build on Failure` option, to enable the functionality.

### CODE REVIEW

File: src/main/resources/net/masterthought/jenkins/CucumberReportPublisher/config.jelly [hunk 1]
```diff
@@ -84,6 +84,13 @@
                     description="${%buildStatus.description}">
                 <f:select/>
             </f:entry>
+
+            <f:entry
+                    title="${%stopBuildOnFailure.title}"
+                    field="stopBuildOnFailure"
+                    description="${%stopBuildOnFailure.description}">
+                <f:checkbox/>
+            </f:entry>
```
[the remark this answers is not in the document]
Comment (katsadim) [reply src/main/resources/net/masterthought/jenkins/CucumberReportPublisher/config.jelly]: I uploaded a new screenshot of advanced options section. Please let me know if it looks ok cause it does not look that good from my laptop screen.

File: src/main/resources/net/masterthought/jenkins/Messages.properties [hunk 2]
```diff
@@ -13,4 +13,6 @@ ReducingMethod.MERGE_FEATURES_BY_ID=Merge features with the same ID
 # ===
 Configuration.notValidNumber=Provide integer value
 Configuration.notValidNumberRange=Provide positive value
+# ===
+StopOnBuildFailure.FailNote=Cucumber tests failed
```
[the remark this answers is not in the document]
Comment (katsadim) [reply src/main/resources/net/masterthought/jenkins/Messages.properties]: done

File: src/main/java/net/masterthought/jenkins/CucumberReportPublisher.java [hunk 3]
```diff
@@ -55,6 +56,7 @@
     private int failedScenariosNumber;
     private int failedFeaturesNumber;
     private String buildStatus;
+    private boolean stopBuildOnFailure = false;
```
[the remark this answers is not in the document]
Comment (katsadim) [reply src/main/java/net/masterthought/jenkins/CucumberReportPublisher.java]: removed

File: src/main/java/net/masterthought/jenkins/CucumberReportPublisher.java [hunk 4]
```diff
@@ -299,6 +310,10 @@ private void generateReport(Run<?, ?> build, FilePath workspace, TaskListener li
         Reportable result = reportBuilder.generateReports();
 
         if (hasReportFailed(result, listener)) {
+            if(stopBuildOnFailure) {
```
[the remark this answers is not in the document]
Comment (katsadim) [reply src/main/java/net/masterthought/jenkins/CucumberReportPublisher.java]: fixed

File: src/main/resources/net/masterthought/jenkins/CucumberReportPublisher/help-stopBuildOnFailure.html [hunk 5]
```diff
@@ -0,0 +1,2 @@
+<p>The default behaviour is to carry on with the build even if the cucumber report contains failures. This option stops the build.
+    Particularly useful when for example there is a need for all tests to pass before deploying to production.</p>
```
[the remark this answers is not in the document]
Comment (katsadim) [reply src/main/resources/net/masterthought/jenkins/CucumberReportPublisher/help-stopBuildOnFailure.html]: newline added

File: src/main/java/net/masterthought/jenkins/CucumberReportPublisher.java [hunk 6]
```diff
@@ -299,6 +310,10 @@ private void generateReport(Run<?, ?> build, FilePath workspace, TaskListener li
...
+                throw new AbortException(Messages.StopOnBuildFailure_FailNote());
```
[the remark this answers is not in the document]
Comment (katsadim) [reply src/main/java/net/masterthought/jenkins/CucumberReportPublisher.java]: Well spotted. I moves this bit below the code that defines the build status.

File: src/main/java/net/masterthought/jenkins/CucumberReportPublisher.java [hunk 7]
```diff
@@ -194,6 +196,15 @@ public void setBuildStatus(String buildStatus) {
         this.buildStatus = buildStatus;
     }
 
+    @DataBoundSetter
+    public void setStopBuildOnFailure(boolean stopBuildOnFailure) {
+        this.stopBuildOnFailure = stopBuildOnFailure;
+    }
+
+    public boolean getStopBuildOnFailure() {
```
[the remark this answers is not in the document]
Comment (katsadim) [reply src/main/java/net/masterthought/jenkins/CucumberReportPublisher.java]: oops sorry i forgot to rebase. How should I proceed now?

File: src/main/java/net/masterthought/jenkins/CucumberReportPublisher.java [hunk 8]
```diff
@@ -225,6 +235,8 @@ public String getClassificationsFilePattern() {
     public void perform(@Nonnull Run<?, ?> run, @Nonnull FilePath workspace, @Nonnull Launcher launcher, @Nonnull TaskListener listener)
             throws InterruptedException, IOException {
 
+        keepBackwardCompatibility();
```
Comment (damianszczepanik) [inline src/main/java/net/masterthought/jenkins/CucumberReportPublisher.java]: this line has been already added https://github.com/jenkinsci/cucumber-reports-plugin/commit/5666fa2ef2e8d76aa350e472a9a82abba336381a#diff-848f2235483f983acf53416159f60a9dR227 so this should not be present here after rebase operation

**Pull request closed without merging**